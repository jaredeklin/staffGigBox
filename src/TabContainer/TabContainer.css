
.tab-container {
  margin: 65px auto;
  display: flex;
  width: 80%;
  height: 500px;
  overflow-y: scroll;
}

ul {
  margin: 0px;
  padding: 0px;
  width: 15%;
}

.tab {
  background: #6161a7;
  color: white;
  height: 75px;
  display: block;
  text-align: center;
  cursor: pointer;
  border-bottom: 1px solid white;
  padding-top: 30px;
  box-sizing: border-box;
}

.tab.active {
  background: #969696;
}

.active-content {
  width: 85%;
  padding: 20px;
  border: 1px solid black;
  overflow-y: scroll;
}