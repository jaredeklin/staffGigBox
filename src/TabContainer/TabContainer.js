import React, { Component } from 'react'
import { Tab } from '../Tab/Tab'
import { EventForm } from '../EventForm/EventForm'
import { StaffForm } from '../StaffForm/StaffForm'
import { Schedule } from '../Schedule/Schedule'
import { Api } from '../Api/Api'
import './TabContainer.css'

export class TabContainer extends Component {

  constructor(props) {
    super(props);
    this.api = new Api()
    this.state = {
      manualSchedule: false,
      manualScheduleData: {},
      activeTabName: this.props.tabs[0]
    };
  }

  handleTabClick = (activeTabName) => this.setState({ activeTabName });

  checkManualSchedule = (eventData, manualSchedule) => {
    this.setState({ 
      manualSchedule,
      manualScheduleData: eventData 
    })
    console.log(manualSchedule)
    if ( !manualSchedule ) {
      this.props.scheduleGenerator()
    }
  }

  updateSchedule = async (event_id) => {
    const manualScheduleData = await this.api.getSchedule(event_id)
    this.setState({ manualScheduleData })
  }

  displayTabs = () => {
    const { activeTabName } = this.state

    return this.props.tabs.map((tabName, index) => {
      return (
        <Tab
          tabName={ tabName }
          handleTabClick={ this.handleTabClick }
          isActive={ tabName === activeTabName }
          key={ tabName + index }
        />
      )
    })
  };

  checkForManual = () => {
    if ( this.state.manualSchedule ) {
      return (
        <Schedule 
          staffList={ this.props.staff }
          editSchedule={ this.props.editSchedule } 
          event={ this.state.manualScheduleData }
          manualSchedule={ true }
          updateSchedule={ this.updateSchedule}
          admin={ this.props.admin }
        />)
    } else {
      return (
        <EventForm 
          scheduleGenerator={ this.props.scheduleGenerator }
          checkManualSchedule={ this.checkManualSchedule } 
        />
      );
    }
  }

  activeContent = () => {
    switch (this.state.activeTabName) {
      case 'Add Event': return this.checkForManual()

      case 'Add New Staff': return <StaffForm addStaff={ this.props.addStaff } user={ this.props.user }/>;

      default: return this.props.schedule.map((event, index) => {
        return (
          <Schedule
            editSchedule={ this.props.editSchedule }
            staffList= { this.props.staff }
            event={ event }
            key={ event.event_id }
            deleteFromSchedule={this.props.deleteFromSchedule}
            admin={ this.props.admin }
          />
        )
      })
    }
  }

  render() {
    return (
      <section className="tab-container">
        <ul>
          { this.displayTabs() }
        </ul>
        <article className="active-content">
          { this.activeContent() }
        </article>
      </section>
    );
  }
};
