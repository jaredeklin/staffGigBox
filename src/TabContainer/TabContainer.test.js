import React from 'react';
import { shallow } from 'enzyme';
import { TabContainer } from './TabContainer';

describe('TabContainer', () => {
  let wrapper;
  let mockEditSchedule = jest.fn();
  let mockTabs = ['Schedule'];
  let mockSchedule = [{ event_id: 1 }];
  let mockScheduleGenerator = jest.fn();

  beforeEach(() => {
    wrapper = shallow(
      <TabContainer
        editSchedule={mockEditSchedule}
        tabs={mockTabs}
        schedule={mockSchedule}
        scheduleGenerator={mockScheduleGenerator}
      />
    );
  });
  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the snapshot with admin tabs', () => {
    const mockAdminTabs = [
      'Add Event',
      'Add New Staff',
      'Schedule',
      'Submit Availability'
    ];

    wrapper = shallow(
      <TabContainer
        editSchedule={mockEditSchedule}
        tabs={mockAdminTabs}
        schedule={mockSchedule}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should match the snapshot when submit Availability is the active tab', () => {
    wrapper.setState({ activeTabName: 'Submit Availability' });
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the snapshot when Add New Staff is the active tab', () => {
    wrapper.setState({ activeTabName: 'Add New Staff' });
    expect(wrapper).toMatchSnapshot();
  });

  it('should set state when handleTabClick is called', () => {
    wrapper.instance().handleTabClick('Add Event');
    expect(wrapper.state('activeTabName')).toEqual('Add Event');
  });

  describe('checkManualSchedule', () => {
    it('should update state and call scheduleGenerator', () => {
      const mockEventData = { event_id: 2 };
      const mockManualSchedule = false;

      wrapper.instance().checkManualSchedule(mockEventData, mockManualSchedule);

      expect(wrapper.state('manualSchedule')).toEqual(false);
      expect(mockScheduleGenerator).toHaveBeenCalled();
    });

    it('should update state when manualSchedule is truthy', () => {
      const mockEventData = { event_id: 2 };
      const mockManualSchedule = true;

      wrapper.instance().checkManualSchedule(mockEventData, mockManualSchedule);

      expect(wrapper.state('manualSchedule')).toEqual(true);
      expect(wrapper.state('manualScheduleData')).toEqual(mockEventData);
    });
  });

  describe('updateSchedule', () => {
    it('should call getSchedule and update state', async () => {
      wrapper.instance().api.getSchedule = jest
        .fn()
        .mockReturnValue({ event_id: 2 });

      await wrapper.instance().updateSchedule(1);

      expect(wrapper.instance().api.getSchedule).toHaveBeenCalledWith(1);
      expect(wrapper.state('manualScheduleData')).toEqual({ event_id: 2 });
    });
  });

  describe('checkForManual', () => {
    it('should match the snapshot when manualSchedule state is true', () => {
      wrapper.setState({ manualSchedule: true });

      wrapper.instance().checkForManual();
      expect(wrapper).toMatchSnapshot();
    });
  });
});
