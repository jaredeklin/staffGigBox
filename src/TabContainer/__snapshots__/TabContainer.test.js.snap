// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabContainer should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TabContainer
    editSchedule={[Function]}
    schedule={
        Array [
            Object {
              "event_id": 1,
            },
          ]
    }
    scheduleGenerator={[Function]}
    tabs={
        Array [
            "Schedule",
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <ul>
          <Unknown
                    handleTabClick={[Function]}
                    isActive={true}
                    tabName="Schedule"
          />
</ul>,
        <article
          className="active-content"
>
          <Schedule
                    admin={undefined}
                    deleteFromSchedule={undefined}
                    editSchedule={[Function]}
                    event={
                              Object {
                                        "event_id": 1,
                                      }
                    }
                    staffList={undefined}
          />
</article>,
      ],
      "className": "tab-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Unknown
              handleTabClick={[Function]}
              isActive={true}
              tabName="Schedule"
/>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "Schedule0",
            "nodeType": "function",
            "props": Object {
              "handleTabClick": [Function],
              "isActive": true,
              "tabName": "Schedule",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "ul",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Schedule
              admin={undefined}
              deleteFromSchedule={undefined}
              editSchedule={[Function]}
              event={
                            Object {
                                          "event_id": 1,
                                        }
              }
              staffList={undefined}
/>,
          ],
          "className": "active-content",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "class",
            "props": Object {
              "admin": undefined,
              "deleteFromSchedule": undefined,
              "editSchedule": [Function],
              "event": Object {
                "event_id": 1,
              },
              "staffList": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "article",
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <ul>
            <Unknown
                        handleTabClick={[Function]}
                        isActive={true}
                        tabName="Schedule"
            />
</ul>,
          <article
            className="active-content"
>
            <Schedule
                        admin={undefined}
                        deleteFromSchedule={undefined}
                        editSchedule={[Function]}
                        event={
                                    Object {
                                                "event_id": 1,
                                              }
                        }
                        staffList={undefined}
            />
</article>,
        ],
        "className": "tab-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Unknown
                handleTabClick={[Function]}
                isActive={true}
                tabName="Schedule"
/>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "Schedule0",
              "nodeType": "function",
              "props": Object {
                "handleTabClick": [Function],
                "isActive": true,
                "tabName": "Schedule",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "ul",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Schedule
                admin={undefined}
                deleteFromSchedule={undefined}
                editSchedule={[Function]}
                event={
                                Object {
                                                "event_id": 1,
                                              }
                }
                staffList={undefined}
/>,
            ],
            "className": "active-content",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "class",
              "props": Object {
                "admin": undefined,
                "deleteFromSchedule": undefined,
                "editSchedule": [Function],
                "event": Object {
                  "event_id": 1,
                },
                "staffList": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "article",
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`TabContainer should match the snapshot with admin tabs 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TabContainer
    editSchedule={[Function]}
    schedule={
        Array [
            Object {
              "event_id": 1,
            },
          ]
    }
    tabs={
        Array [
            "Add Event",
            "Add New Staff",
            "Schedule",
            "Submit Availability",
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <ul>
          <Unknown
                    handleTabClick={[Function]}
                    isActive={true}
                    tabName="Add Event"
          />
          <Unknown
                    handleTabClick={[Function]}
                    isActive={false}
                    tabName="Add New Staff"
          />
          <Unknown
                    handleTabClick={[Function]}
                    isActive={false}
                    tabName="Schedule"
          />
          <Unknown
                    handleTabClick={[Function]}
                    isActive={false}
                    tabName="Submit Availability"
          />
</ul>,
        <article
          className="active-content"
>
          <EventForm
                    checkManualSchedule={[Function]}
                    scheduleGenerator={undefined}
          />
</article>,
      ],
      "className": "tab-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Unknown
              handleTabClick={[Function]}
              isActive={true}
              tabName="Add Event"
/>,
            <Unknown
              handleTabClick={[Function]}
              isActive={false}
              tabName="Add New Staff"
/>,
            <Unknown
              handleTabClick={[Function]}
              isActive={false}
              tabName="Schedule"
/>,
            <Unknown
              handleTabClick={[Function]}
              isActive={false}
              tabName="Submit Availability"
/>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "Add Event0",
            "nodeType": "function",
            "props": Object {
              "handleTabClick": [Function],
              "isActive": true,
              "tabName": "Add Event",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "Add New Staff1",
            "nodeType": "function",
            "props": Object {
              "handleTabClick": [Function],
              "isActive": false,
              "tabName": "Add New Staff",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "Schedule2",
            "nodeType": "function",
            "props": Object {
              "handleTabClick": [Function],
              "isActive": false,
              "tabName": "Schedule",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "Submit Availability3",
            "nodeType": "function",
            "props": Object {
              "handleTabClick": [Function],
              "isActive": false,
              "tabName": "Submit Availability",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "ul",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <EventForm
            checkManualSchedule={[Function]}
            scheduleGenerator={undefined}
/>,
          "className": "active-content",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "checkManualSchedule": [Function],
            "scheduleGenerator": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "article",
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <ul>
            <Unknown
                        handleTabClick={[Function]}
                        isActive={true}
                        tabName="Add Event"
            />
            <Unknown
                        handleTabClick={[Function]}
                        isActive={false}
                        tabName="Add New Staff"
            />
            <Unknown
                        handleTabClick={[Function]}
                        isActive={false}
                        tabName="Schedule"
            />
            <Unknown
                        handleTabClick={[Function]}
                        isActive={false}
                        tabName="Submit Availability"
            />
</ul>,
          <article
            className="active-content"
>
            <EventForm
                        checkManualSchedule={[Function]}
                        scheduleGenerator={undefined}
            />
</article>,
        ],
        "className": "tab-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Unknown
                handleTabClick={[Function]}
                isActive={true}
                tabName="Add Event"
/>,
              <Unknown
                handleTabClick={[Function]}
                isActive={false}
                tabName="Add New Staff"
/>,
              <Unknown
                handleTabClick={[Function]}
                isActive={false}
                tabName="Schedule"
/>,
              <Unknown
                handleTabClick={[Function]}
                isActive={false}
                tabName="Submit Availability"
/>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "Add Event0",
              "nodeType": "function",
              "props": Object {
                "handleTabClick": [Function],
                "isActive": true,
                "tabName": "Add Event",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "Add New Staff1",
              "nodeType": "function",
              "props": Object {
                "handleTabClick": [Function],
                "isActive": false,
                "tabName": "Add New Staff",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "Schedule2",
              "nodeType": "function",
              "props": Object {
                "handleTabClick": [Function],
                "isActive": false,
                "tabName": "Schedule",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "Submit Availability3",
              "nodeType": "function",
              "props": Object {
                "handleTabClick": [Function],
                "isActive": false,
                "tabName": "Submit Availability",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "ul",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <EventForm
              checkManualSchedule={[Function]}
              scheduleGenerator={undefined}
/>,
            "className": "active-content",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "checkManualSchedule": [Function],
              "scheduleGenerator": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "article",
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
